packages/flutter/lib/src/painting/text_painter.dart

```diff
@@ -630,15 +662,23 @@ class TextPainter {
   /// To set the text style, specify a [TextStyle] when creating the [TextSpan]
   /// that you pass to the [TextPainter] constructor or to the [text] property.
   void paint(Canvas canvas, Offset offset) {
-    assert(() {
-      if (_needsLayout) {
-        throw FlutterError(
-          'TextPainter.paint called when text geometry was not yet calculated.\n'
-          'Please call layout() before paint() to position the text before painting it.',
-        );
-      }
-      return true;
-    }());
+    final double? minWidth = _lastMinWidth;
+    final double? maxWidth = _lastMaxWidth;
+    if (_paragraph == null || minWidth == null || maxWidth == null) {
+      throw StateError(
+        'TextPainter.paint called when text geometry was not yet calculated.\n'
+        'Please call layout() before paint() to position the text before painting it.',
+      );
+    }
+
+    if (_needsPaint) {
+      _createParagraph();
+      // Unfortunately we have to redo the layout using the same constraints,
+      // since we've created a new ui.Paragraph. But there's no extra work being
+      // done: if _needsPaint is true and _paragraph is not null, the previous
+      // `layout` call didn't invoke _layoutParagraph.
+      _layoutParagraph(minWidth, maxWidth);
```

Comment:
Can we assert that the size calculated by the new paragraph is the same as the size before?